src/js/common/containers/DowntimeNotification.jsx

```diff
@@ -0,0 +1,248 @@
+    determineStatus: PropTypes.func,
+    getScheduledDowntime: PropTypes.func.isRequired,
+    isReady: PropTypes.bool,
+    loadingIndicator: PropTypes.node,
+    render: PropTypes.func,
+    scheduledDowntime: PropTypes.arrayOf(
+      PropTypes.shape({
+        service: PropTypes.string,
+        description: PropTypes.string,
+        startTime: PropTypes.instanceOf(Date),
+        endTime: PropTypes.instanceOf(Date)
+      })
+    ),
+    userIsAuthenticated: PropTypes.bool
+  };
+
+  static defaultProps = {
+    appTitle: 'application'
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      modalDismissed: false,
+      cache: {}
+    };
+  }
+
+  componentWillMount() {
+    this.props.getScheduledDowntime();
+  }
+
+  // This is here just for caching the result of calculateDowntime, so that it isn't run every time a prop changes.
+  // Currently, this component should only do its calculations once, because it would be weird if in the middle of filling
+  // out a form a modal or alert appears alerting them of downtime, or if it just shuts down the app because we entered
+  // scheduled downtime.
+  componentWillReceiveProps(newProps) {
+    const firstLoad = newProps.isReady && !this.props.isReady;
+    if (firstLoad) {
+      const downtimeMap = this.calculateDowntime(newProps.dependencies, newProps.scheduledDowntime);
+      const status = this.determineStatus(downtimeMap);
+      const downtimeWindow = this.getDowntimeWindow(downtimeMap[status]);
+      const cache = { downtimeMap, status, downtimeWindow };
+      this.setState({ cache });
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    if (!nextProps.isReady) return false;
+    if (nextProps.isReady && !this.props.isReady) return true;
+
+    return nextState.cache.status !== serviceStatus.down;
+  }
+
+  getStatusForDowntime(downtime, current = moment(), warning = moment().add(1, 'hour')) {
+    const inclusive = '[]';
+
+    if (!downtime) return serviceStatus.ok;
+    if (current.isBetween(downtime.startTime, downtime.endTime, inclusive)) return serviceStatus.down;
+    if (warning.isBetween(downtime.startTime, downtime.endTime, inclusive)) return serviceStatus.downtimeApproaching;
+
+    return serviceStatus.ok;
+  }
+
+  getDowntimeWindow(downtimes) {
+    return downtimes.reduce((result, downtime) => {
+      if (!downtime.startTime || !downtime.endTime) return result;
+      const startTime = moment(downtime.startTime);
+      const endTime = moment(downtime.endTime);
+      return {
+        startTime: result.startTime && result.startTime.isBefore(startTime) ? result.startTime : startTime,
+        endTime: result.endTime && result.endTime.isAfter(endTime) ? result.endTime : endTime,
+      };
+    }, {});
+  }
+
+  // Converts the array of dependencies/service names into key/value pairs, with service statuses as keys and a list of
+  // downtime information (each downtime corresponding to a dependency/service name) as the values.
```

Comment:
Perhaps an example would help clear this description up? Could just be me, but I was having a hard time making sense of it.

Reply:
Good idea, will do